### PR TITLE
chore: reward refactor and test

### PR DESCRIPTION
**Pull request by u2**
Why I put the two commits in the same PR?

The reason is that without the refactoring, the last test case would be more complicated than now, it would need more data to craft.

There is no cache in store, we should avoid DB query. Later I will try to add some cache in the store or somewhere.

### CODE REVIEW

File: util/reward-calculator/src/lib.rs [hunk 1]
```diff
@@ -199,3 +215,375 @@ impl<'a, P: ChainProvider> RewardCalculator<'a, P> {
...
+    //
+    // pn denotes poposal
+    // block-10: p1
+    // block-11: p2, uncles-proposals: p3
+    // block-13 [target]: p1, p3, p4, p5, uncles-proposals: p6
+    // block-14: p4, txs(p1, p2, p3)
+    // block-15: txs(p4)
+    // block-18: txs(p5, p6)
+    // block-19 [current]
+    // target's earliest proposer: p4, p5, p6
```
Comment (u2) [inline util/reward-calculator/src/lib.rs]: should be `earliest proposals`,  change it later.

File: util/reward-calculator/src/lib.rs [hunk 2]
```diff
@@ -199,3 +215,375 @@ impl<'a, P: ChainProvider> RewardCalculator<'a, P> {
...
+            store: Arc::new(store),
+            consensus,
+        };
+
+        let reward_calculator = RewardCalculator::new(&provider);
+        let proposal_reward = reward_calculator
+            .proposal_reward(block_18.header(), block_13.header())
+            .unwrap();
+
+        // target's earliest proposer: p4, p5, p6
```
Comment (u2) [inline util/reward-calculator/src/lib.rs]: should be `proposals`

**Comment by xxuejie** [comment]
Notice the rules here will change once #1094 lands

**Comment by u2** [comment]
> Notice the rules tested here will change once #1094 lands

Thanks, I will review the PR.

**Comment by doitian** [comment]
#1094 merged

**Comment by u2** [comment]
> #1094 merged

rebased.

File: util/reward-calculator/src/lib.rs [hunk 3]
```diff
@@ -122,23 +123,42 @@ impl<'a, P: ChainProvider> RewardCalculator<'a, P> {
...
+        let txs_fees_proc = |hash: &H256| -> Vec<Capacity> {
+            store
+                .get_block_ext(hash)
+                .expect("block ext stored")
+                .txs_fees
+        };
+
+        let has_committed_proc =
+            |x: &FnvHashSet<ProposalShortId>, y: &FnvHashSet<ProposalShortId>| -> bool {
+                !x.intersection(&y).collect::<FnvHashSet<_>>().is_empty()
```
[the remark this answers is not in the document]
Comment (u2) [reply util/reward-calculator/src/lib.rs]: updated

**Pull request merged**